ciao-controller/internal/datastore/datastore.go

```diff
@@ -1682,3 +1725,511 @@ func (ds *Datastore) GetVolumeAttachments(volume string) ([]types.StorageAttachm
 
 	return attachments, nil
 }
+
+// GetPool will return an external IP Pool
+func (ds *Datastore) GetPool(ID string) (types.Pool, error) {
+	ds.poolsLock.RLock()
+	p, ok := ds.pools[ID]
+	ds.poolsLock.RUnlock()
+
+	if !ok {
+		return p, types.ErrPoolNotFound
+	}
+
+	return p, nil
+}
+
+// GetPools will return a list of external IP Pools
+func (ds *Datastore) GetPools() ([]types.Pool, error) {
+	var pools []types.Pool
+
+	ds.poolsLock.RLock()
+
+	for _, p := range ds.pools {
+		pools = append(pools, p)
+	}
+
+	ds.poolsLock.RUnlock()
+
+	return pools, nil
+}
+
+// lock for the map must be held by caller.
+func (ds *Datastore) isDuplicateSubnet(new *net.IPNet) bool {
+	for s, exists := range ds.externalSubnets {
+		if exists == true {
+			// this will always succeed
+			_, subnet, _ := net.ParseCIDR(s)
+
+			if subnet.Contains(new.IP) || new.Contains(subnet.IP) {
+				return true
+			}
+		}
+	}
+
+	return false
+}
+
+// lock for the map must be held by the caller
+func (ds *Datastore) isDuplicateIP(new net.IP) bool {
+	// first make sure the IP isn't covered by a subnet
+	for s, exists := range ds.externalSubnets {
+		// this will always succeed
+		_, subnet, _ := net.ParseCIDR(s)
+
+		if exists == true {
+			if subnet.Contains(new) {
+				return true
+			}
+		}
+	}
+
+	// next make sure that the IP isn't already in a
+	// different pool
+	return ds.externalIPs[new.String()]
+}
+
+// AddPool will add a brand new pool to our datastore.
+func (ds *Datastore) AddPool(pool types.Pool) error {
+	ds.poolsLock.Lock()
+	defer ds.poolsLock.Unlock()
+
+	if len(pool.Subnets) > 0 {
+		// check each one to make sure it's not in use.
+		for _, subnet := range pool.Subnets {
+			_, newSubnet, err := net.ParseCIDR(subnet.CIDR)
+			if err != nil {
+				return err
+			}
+
+			if ds.isDuplicateSubnet(newSubnet) {
+				return types.ErrDuplicateSubnet
+			}
+
+			// update our list of used subnets
+			ds.externalSubnets[subnet.CIDR] = true
+		}
+	} else if len(pool.IPs) > 0 {
+		var newIPs []net.IP
+
+		// make sure valid and not duplicate
+		for _, newIP := range pool.IPs {
+			IP := net.ParseIP(newIP.Address)
+			if IP == nil {
+				return types.ErrInvalidIP
+			}
+
+			if ds.isDuplicateIP(IP) {
+				return types.ErrDuplicateIP
+			}
+
+			newIPs = append(newIPs, IP)
+		}
+
+		// now that the whole list is confirmed, we can update
+		for _, IP := range newIPs {
+			ds.externalIPs[IP.String()] = true
+		}
+	}
+
+	ds.pools[pool.ID] = pool
+	err := ds.db.createPool(pool)
+	if err != nil {
+		ds.poolsLock.Unlock()
```

Comment:
Won't this cause a runtime panic when the defer runs, because the mutex will be unlocked twice?